R/sads-methods.R

```diff
@@ -342,41 +342,62 @@ setMethod("radpred", signature(object="missing", sad="character", rad="missing",
+    oct <- c(octlower, oct)
+  }
+  oct
+}
           
 ## octavpred generic functions and methods ###
 setGeneric("octavpred",
-def = function(object, sad, rad, coef, trunc, oct, S, N, ...) standardGeneric("octavpred"))
+def = function(object, sad, rad, coef, trunc, oct, S, N, preston=FALSE, ...) standardGeneric("octavpred"))
 
 ## if object is of class fitsad (no other argument should be provided)
 setMethod("octavpred", signature(object="fitsad",sad="missing", rad="missing",
                                  coef="missing", trunc="missing", oct="ANY",
                                  S="missing", N="missing"),
-          function(object, sad, rad, coef, trunc, oct, S, N, ...){
-            dots <- list(...)
-            coef <- as.list(bbmle::coef(object))
-            trunc <- object@trunc
-            sad <- object@sad
+          function(object, sad, rad, coef, trunc, oct, S, N, preston, ...){
             x <- object@data$x
-            S <- length(x)
-            N <- sum(x)
-            if(missing(oct)){
-                oct <- 0:(ceiling(max(log2(x)))+1)
-                if(any(x < 1)){
-                    octlower <- floor(min(log2((x)))):-1
-                    oct <- c(octlower, oct)
-                }
-            }
+            if(missing(oct)) oct <- genoct(x)
+            octavpred(sad = object@sad, coef = as.list(bbmle::coef(object)),
+                      trunc = object@trunc, oct = oct, S=length(x), N=sum(x), preston=preston, ...)
+          }
+          )
+## if object is a numeric vector of abundances and sad argument is given (rad, S, N,  arguments should be missing)
+setMethod("octavpred", signature(object="numeric",sad="character", rad="missing",
+                                 coef="list", oct="ANY", trunc="ANY", S="missing", N="missing"),
+          function(object, sad, rad, coef, trunc, oct, S, N, preston, ...){
+            if(missing(oct)) oct <- genoct(object)
+            if(missing(trunc)) trunc<-NaN
+            octavpred(sad=sad, coef=coef, trunc=trunc, oct=oct, S = length(object), N = sum(object),
+                      preston=preston, ...)
+          }
+          )
+## Octavpred workhorse for "sads"
+setMethod("octavpred", signature(object="missing",sad="character", rad="missing",
+                                 coef="list", trunc="ANY", oct="numeric", S="numeric", N="numeric"),
+          function(object, sad, rad, coef, trunc, oct, S, N, preston, ...){
+            dots <- list(...)
             oct <- unique(oct)
-            n <- 2^oct
-            if(!is.na(trunc)){
-				Y <- do.call(ptrunc, c(list(sad, q = n, coef = coef, trunc = trunc), dots))
-            }
-            else{
-				psad <- get(paste("p",sad,sep=""),mode="function")
-				Y <- do.call(psad, c(list(q = n),coef,dots))
+            if (preston) {
+              if(missing(trunc)) trunc <- NaN
+              return(octav(radpred(sad=sad, coef=coef, trunc=trunc, distr="C", S=S, N=N)$abund, preston=TRUE))
```

Comment:
I think that this should be distr="D" to work with discrete distributions. Not important anyway because distr slot is deprecated by PR #86